README.md
# NAME

    Data::Tools provides set of basic functions for data manipulation.

# SYNOPSIS

    use Data::Tools qw( :all );  # import all functions
    use Data::Tools;             # the same as :all :) 
    use Data::Tools qw( :none ); # do not import anything, use full package names

    # --------------------------------------------------------------------------

    data_tools_set_file_io_encoding( 'UTF-8' ); # all file IO will use UTF-8
    data_tools_set_file_io_encoding( ':RAW' );  # all file IO will use binary data

    my $res  = file_save( $file_name, 'file content here' );
    my $data = file_load( $file_name );

    my $data_arrayref = file_load_ar( $file_name );
    
    # for specific charset encoding and because of backward compatibility:

    my $res  = file_save( { FILE_NAME => $file_name, ENCODING => 'UTF-8' }, 'data' );
    my $data = file_load( { FILE_NAME => $file_name, ENCODING => 'UTF-8' } );

    my $data_arrayref = file_load_ar( { FILE_NAME => $fname, ENCODING => 'UTF-8' } );

    # --------------------------------------------------------------------------

    my $file_modification_time_in_seconds = file_mtime( $file_name );
    my $file_change_time_in_seconds       = file_ctime( $file_name );
    my $file_last_access_time_in_seconds  = file_atime( $file_name );
    my $file_size                         = file_size(  $file_name );

    # --------------------------------------------------------------------------
    
    my $res  = dir_path_make( '/path/to/somewhere' ); # create full path with 0700
    my $res  = dir_path_make( '/new/path', MASK => 0755 ); # ...with mask 0755
    my $path = dir_path_ensure( '/path/s/t/h' ); # ensure path exists, check+make

    # --------------------------------------------------------------------------

    my $path_with_trailing_slash = file_path( $full_path_or_file_name );

    # file_name() and file_name_ext() return full name with leadeing 
    # dot for dot-files ( .filename )
    my $file_name_including_ext  = file_name_ext( $full_path_or_file_name );
    my $file_name_only_no_ext    = file_name( $full_path_or_file_name );

    # file_ext() returns undef for dot-files ( .filename )
    my $file_ext_only            = file_ext( $full_path_or_file_name );

    # --------------------------------------------------------------------------
    
    # uses simple backslash escaping of \n, = and \ itself
    my $data_str = hash2str( $hash_ref ); # convert hash to string "key=value\n"
    my $hash_ref = str2hash( $hash_str ); # convert str "key-value\n" to hash

    # same as hash2str() but uses keys in certain order
    my $data_str = hash2str_keys( \%hash, sort keys %hash );
    my $data_str = hash2str_keys( \%hash, sort { $a <=> $b } keys %hash );

    # same as hash2str() and str2hash() but uses URL-style escaping
    my $data_str = hash2str_url( $hash_ref ); # convert hash to string "key=value\n"
    my $hash_ref = str2hash_url( $hash_str ); # convert str "key-value\n" to hash
    
    my $hash_ref = url2hash( 'key1=val1&key2=val2&testing=tralala);
    # $hash_ref will be { key1 => 'val1', key2 => 'val2', testing => 'tralala' }

    my $hash_ref_with_upper_case_keys = hash_uc( $hash_ref_with_lower_case_keys );
    my $hash_ref_with_lower_case_keys = hash_lc( $hash_ref_with_upper_case_keys );

    hash_uc_ipl( $hash_ref_to_be_converted_to_upper_case_keys );
    hash_lc_ipl( $hash_ref_to_be_converted_to_lower_case_keys );
    
    # save/load hash in str_url_escaped form to/from a file
    my $res      = hash_save( $file_name, $hash_ref );
    my $hash_ref = hash_load( $file_name );

    # save hash with certain keys order, uses hash2str_keys()
    my $res      = hash_save( $file_name, \%hash, sort keys %hash );
    
    # same as hash_save() and hash_load() but uses hash2str_url() and str2hash_url()
    my $res      = hash_save_url( $file_name, $hash_ref );
    my $hash_ref = hash_load_url( $file_name );

    # validate (nested) hash by example
    
    # validation example nested hash
    my $validate_hr = {
                      A => 'INT',
                      B => 'INT(-5,10)',
                      C => 'REAL',
                      D => {
                           E => 'RE:\d+[a-f]*',  # regexp match
                           F => 'REI:\d+[a-f]*', # case insensitive regexp match
                           },
                      DIR1  => '-d',   # must be existing directory
                      DIR2  => 'dir',  # must be existing directory
                      FILE1 => '-f',   # must be existing file  
                      FILE2 => 'file', # must be existing file  
                      };
    # actual nested hash to be verified if looks like the example
    my $data_hr     = {
                      A => '123',
                      B =>  '-1',
                      C =>  '1 234 567.89',
                      D => {
                           E => '123abc',
                           F => '456FFF',
                           },
                      }               
    
    my @invalid_keys = hash_validate( $data_hr, $validate_hr );
    print "YES!" if hash_validate( $data_hr, $validate_hr );

    # --------------------------------------------------------------------------
    
    my $escaped   = str_url_escape( $plain_str ); # URL-style %XX escaping
    my $plain_str = str_url_unescape( $escaped );

    my $escaped   = str_html_escape( $plain_str ); # HTML-style &name; escaping
    my $plain_str = str_html_unescape( $escaped );
    
    my $hex_str   = str_hex( $plain_str ); # HEX-style XX string escaping
    my $plain_str = str_unhex( $hex_str );

    # --------------------------------------------------------------------------
    
    # converts perl package names to file names, f.e: returns "Data/Tools.pm"
    my $perl_pkg_fn = perl_package_to_file( 'Data::Tools' );

    # --------------------------------------------------------------------------

    # calculating hex digests
    my $whirlpool_hex = wp_hex( $data );
    my $sha1_hex      = sha1_hex( $data );
    my $md5_hex       = md5_hex( $data );

    # --------------------------------------------------------------------------

    my $formatted_str = str_num_comma( 1234567.89 );   # returns "1'234'567.89"
    my $formatted_str = str_num_comma( 4325678, '_' ); # returns "4_325_678"
    my $padded_str    = str_pad( 'right', -12, '*' ); # returns "right*******"
    my $str_c         = str_countable( $dc, 'day', 'days' );
                        # returns 'days' for $dc == 0
                        # returns 'day'  for $dc == 1
                        # returns 'days' for $dc >  1

    # --------------------------------------------------------------------------

    # find all *.txt files in all subdirectories starting from /usr/local
    # returned files are with full path names
    my @files = glob_tree( '/usr/local/*.txt' );

    # read directory entries names (without full paths)
    my @files_and_dirs = read_dir_entries( '/tmp/secret/dir' );

# FUNCTIONS

## hash\_validate( $data\_hr, $validate\_hr );

Return value can be either scalar or array context. In scalar context return
value is true (1) or false (0). In array context it returns list of the invalid
keys (possibly key paths like 'KEY1/KEY2/KEY3'):

    # array context
    my @invalid_keys = hash_validate( $data_hr, $validate_hr );
    
    # scalar context
    print "YES!" if hash_validate( $data_hr, $validate_hr );

# TODO

    (more docs)

# DATA::TOOLS SUB-MODULES

Data::Tools package includes several sub-modules:

    * Data::Tools::Socket (socket I/O processing, TODO: docs)
    * Data::Tools::Time   (time processing)

# REQUIRED MODULES

Data::Tools is designed to be simple, compact and self sufficient. 
However it uses some 3rd party modules:

    * Digest::Whirlpool
    * Digest::MD5
    * Digest::SHA1

# SEE ALSO

For more complex cases of nested hash validation, 
check Data::Validate::Struct module by Thomas Linden, cheers :)

# GITHUB REPOSITORY

    git@github.com:cade-vs/perl-data-tools.git
    
    git clone git://github.com/cade-vs/perl-data-tools.git
    

# AUTHOR

    Vladi Belperchinov-Shabanski "Cade"

    <cade@biscom.net> <cade@datamax.bg> <cade@cpan.org>

    http://cade.datamax.bg
# NAME

    Data::Tools::Socket provides set of socket I/O functions.

# SYNOPSIS

    use Data::Tools::Socket qw( :all );  # import all functions
    use Data::Tools::Socket;             # the same as :all :) 
    use Data::Tools::Socket qw( :none ); # do not import anything, use full package names

    # --------------------------------------------------------------------------

    my $read_res_len  = socket_read(  $socket, $data_ref, $length, $timeout );
    my $write_res_len = socket_write( $socket, $data,     $length, $timeout );
    my $write_res_len = socket_print( $socket, $data, $timeout );

    # --------------------------------------------------------------------------

    my $read_data = socket_read_message(  $socket, $timeout );
    my $write_res = socket_write_message( $socket, $data, $timeout );

    # --------------------------------------------------------------------------

# FUNCTIONS

## socket\_read(  $socket, $data\_ref, $length, $timeout )

Reads $length sized data from the $socket and store it to $data\_ref scalar 
reference.

Returns read length (can be shorter than requested $length);

$timeout is optional, it is in seconds and can be less than 1 second.

## socket\_write( $socket, $data,     $length, $timeout )

Writes $length sized data from $data scalar to the $socket.

Returns write length (can be shorter than requested $length);

$timeout is optional, it is in seconds and can be less than 1 second.

## socket\_print( $socket, $data, $timeout )

Same as socket\_write() but calculates requested length from the $data scalar.

$timeout is optional, it is in seconds and can be less than 1 second.

## socket\_read\_message(  $socket, $timeout )

Reads 32bit network-order integer, which then is used as data size to be read
from the socket (i.e. message = 32bit-integer + data ).

Returns read data or undef for message or network error.

$timeout is optional, it is in seconds and can be less than 1 second.

## socket\_write\_message( $socket, $data, $timeout )

Writes 32bit network-order integer, which is the size of the given $data to be
written to the $socket and then writes the data 
(i.e. message = 32bit-integer + data ).

Returns 1 on success or undef for message or network error.

$timeout is optional, it is in seconds and can be less than 1 second.

# TODO

    * more docs

# REQUIRED MODULES

Data::Tools::Socket uses:

    * IO::Select
    * Time::HiRes

# GITHUB REPOSITORY

    git@github.com:cade-vs/perl-data-tools.git
    
    git clone git://github.com/cade-vs/perl-data-tools.git
    

# AUTHOR

    Vladi Belperchinov-Shabanski "Cade"

    <cade@biscom.net> <cade@datamax.bg> <cade@cpan.org>

    http://cade.datamax.bg
# NAME

    Data::Tools::Time provides set of basic functions for time processing.

# SYNOPSIS

    use Data::Tools::Time qw( :all );  # import all functions
    use Data::Tools::Time;             # the same as :all :) 
    use Data::Tools::Time qw( :none ); # do not import anything

    # --------------------------------------------------------------------------

    my $time_diff_str     = unix_time_diff_in_words( $time1 - $time2 );
    my $time_diff_str_rel = unix_time_diff_in_words_relative( $time1 - $time2 );

    # --------------------------------------------------------------------------
      
    my $date_diff_str     = julian_date_diff_in_words( $date1 - $date2 );
    my $date_diff_str_rel = julian_date_diff_in_words_relative( $date1 - $date2 );

    # --------------------------------------------------------------------------
    
    # gets current julian date, needs Time::JulianDay
    my $jd = local_julian_day( time() );

    # move current julian date to year ago, one month ahead and 2 days ahead
    $jd = julian_date_add_ymd( $jd, -1, 1, 2 );

    # get year, month and day from julian date
    my ( $y, $m, $d ) = julian_date_to_ymd( $jd );

    # get julian date from year, month and day
    $jd = julian_date_from_ymd( $y, $m, $d );

    # move julian date ($jd) to the first day of its current month
    $jd = julian_date_goto_first_dom( $jd );

    # move julian date ($jd) to the last day of its current month
    $jd = julian_date_goto_last_dom( $jd );

    # get day of week for given julian date ( 0 => Mon .. 6 => Sun )
    my $dow = julian_date_get_dow( $jd );
    print( ( qw( Mon Tue Wed Thu Fri Sat Sun ) )[ $dow ] . "\n" );

    # get month days count for the given julian date's month
    my $mdays = julian_date_month_days( $jd );

    # get month days count for the given year and month
    my $mdays = julian_date_month_days_ym( $y, $m );

# FUNCTIONS

## unix\_time\_diff\_in\_words( $unix\_time\_diff )

Returns human-friendly text for the given time difference (in seconds).
This function returns absolute difference text, for relative 
(before/after/ago/in) see unix\_time\_diff\_in\_words\_relative().

## unix\_time\_diff\_in\_words\_relative( $unix\_time\_diff )

Same as unix\_time\_diff\_in\_words() but returns relative text
(i.e. with before/after/ago/in)

## julian\_date\_diff\_in\_words( $julian\_date\_diff );

Returns human-friendly text for the given date difference (in days).
This function returns absolute difference text, for relative 
(before/after/ago/in) see julian\_day\_diff\_in\_words\_relative().

## julian\_date\_diff\_in\_words\_relative( $julian\_date\_diff );

Same as julian\_date\_diff\_in\_words() but returns relative text
(i.e. with before/after/ago/in)

# TODO

    * support for language-dependent wording (before/ago)
    * support for user-defined thresholds (48 hours, 2 months, etc.)

# REQUIRED MODULES

Data::Tools::Time uses:

    * Data::Tools (from the same package)
    * Date::Calc
    * Time::JulianDay

# TEXT TRANSLATION NOTES

time/date difference wording functions does not have translation functions
and return only english text. This is intentional since the goal is to keep
the translation mess away but still allow simple (yet bit strange) 
way to translate the result strings with regexp and language hash:

    my $time_diff_str_rel = unix_time_diff_in_words_relative( $time1 - $time2 );
    
    my %TRANS = (
                'now'       => 'sega',
                'today'     => 'dnes',
                'tomorrow'  => 'utre',
                'yesterday' => 'vchera',
                'in'        => 'sled',
                'before'    => 'predi',
                'year'      => 'godina',
                'years'     => 'godini',
                'month'     => 'mesec',
                'months'    => 'meseca',
                'day'       => 'den',
                'days'      => 'dni',
                'hour'      => 'chas',
                'hours'     => 'chasa',
                'minute'    => 'minuta',
                'minutes'   => 'minuti',
                'second'    => 'sekunda',
                'seconds'   => 'sekundi',
                );
                
    $time_diff_str_rel =~ s/([a-z]+)/$TRANS{ lc $1 } || $1/ge;

I know this is no good for longer sentences but works fine in this case.

# GITHUB REPOSITORY

    git@github.com:cade-vs/perl-data-tools.git
    
    git clone git://github.com/cade-vs/perl-data-tools.git
    

# AUTHOR

    Vladi Belperchinov-Shabanski "Cade"

    <cade@bis.bg> <cade@biscom.net> <cade@datamax.bg> <cade@cpan.org>

    http://cade.datamax.bg
# NAME

    Data::Tools::Math provides set of basic functions for mathematics.

# SYNOPSIS

    use Data::Tools::Math qw( :all );  # import all functions
    use Data::Tools::Math;             # the same as :all :) 
    use Data::Tools::Math qw( :none ); # do not import anything

    # --------------------------------------------------------------------------


    # --------------------------------------------------------------------------

# FUNCTIONS

## num\_round( $number, $precision )

Rounds $number to $precisioun places after the decimal point.

## num\_round\_trunc( $number, $precision )

Same as num\_trunc() but just truncates after the $precision places.

## num\_pow( $number, $exponent )

Returns power of $number by $exponent ( $num \*\* $exp )

# REQUIRED MODULES

Data::Tools::Time uses:

    * Math::BigFloat

# GITHUB REPOSITORY

    git@github.com:cade-vs/perl-data-tools.git
    
    git clone git://github.com/cade-vs/perl-data-tools.git
    

# AUTHOR

    Vladi Belperchinov-Shabanski "Cade"

    <cade@bis.bg> <cade@biscom.net> <cade@datamax.bg> <cade@cpan.org>

    http://cade.datamax.bg


Changes
Revision history for Perl extension Data::Tools

1.24 2020-06-22

    - reimplemented str_hex/unhex to use unpack/pack (of course!)
         thanks to String::HexConvert for reminding me that pack 
         is always here :)

    - added separate binary and text file load/save api. 
    
      new functions:
      
          # always binary data load
          file_bin_load( $file_name )

          # always binary data save
          file_bin_save( $file_name, @data, $data, 'data', 123 )
          
      new functions:
          
          # always text data load
          file_text_load( $file_name )

          # always text data load, returns array ref of read data
          file_text_load( $file_name )

          # always text data save
          file_text_save( $file_name, @data, $data, 'data', 123 )
          
      the old encoding defaults to file load/save:
      
          data_tools_set_file_io_encoding( 'cp1251' );
          data_tools_set_file_io_encoding( 'UTF-8'  );
          data_tools_set_file_io_utf8();
          
       still works but will only relate to file_text_* functions.
       
       NOTE: WARNING!
       currently old file_load()/file_save() will default to the old
       behaviour but will be switched to binary only!
       
       NOTE: WARNING!
       file_text_* api will have second optional hashref with operation
       options. for saving, it is uncommon to provide hashref so it will
       be used as options if given.
       
          

1.20 2019-01-18

    - added julian_date_* functions to Data::Tools::Time

1.19 2018-08-31

    - changed escaping style of hash2str() and str2hash()
    - old style, url-escaping are renamed as hash2str_url() and str2hash_url()
    - added hash2str_keys() to allow to specify certain keys order, check docs
    - added hash_save_url() and hash_load_url()
    - added hash_save_keys() to use hash2str_keys(), check docs

1.18 2018-08-10

    - added data_tools_set_file_io_encoding()
    - fixed various file IO encoding issues

1.17 2018-08-07

    - str_num_comma() can accept second argument--padding symbol or string

1.16 2018-07-19

    - added Data::Tools::Math functions
    - added file_load_ar()

1.15 2018-03-31

    - file name path functions added
    - socket read len fixed

1.14 2016-11-30

    - added Data::Tools::Time utilities
    - added Data::Tools::Socket docs
    - added Data::Tools::Time test t/t1-time-diff.pl

1.12 2016-09-16

    - fixed type in Data::Tools::Socket

1.11 2016-09-05 CADE

    - added Data::Tools::Socket utilities

1.10 2016-01-27 CADE

    - fixed non-obvious iterator behaviour of File::Glob inside glob_tree()

1.09 2016-01-03 CADE

    - added:
        hash_keys_lock_recursive

1.08 2015-09-30 CADE

    - added:
        file_mtime  
        file_ctime  
        file_atime  
        file_size  

1.07 2015-09-24 CADE

    - added hash_lock_recursive()
            hash_unlock_recursive()
      to handle missing functionality until perl 5.22 etc.
    - moved to cpan changelog syntax  
    - added glob_tree()

1.06 2013-07-03 CADE
  
    - added more features to hash_validate()
    - changed version to reupload to CPAN

1.05 2013-07-03 CADE
  
    - hash_validate()
    - more docs, "required modules", "see also"

1.04  2013-05-24 CADE

    - changed github urls

1.03  2013-05-04 CADE
  
    - added hash_uc()/hash_lc()

1.02  2013-03-24 CADE

    - added digest funcs
    - fixed docs
    - added github info

1.01  2013-03-07 CADE

    - fixed names in the docs

1.00  2013-03-06 CADE

    - history lost, initial CPAN release

    
    


lib/Data/Tools/Time.pm
##############################################################################
#
#  Data::Tools perl module
#  2013-2019 (c) Vladi Belperchinov-Shabanski "Cade"
#  http://cade.datamax.bg
#  <cade@bis.bg> <cade@biscom.net> <cade@datamax.bg> <cade@cpan.org>
#
#  GPL
#
##############################################################################
package Data::Tools::Time;
use strict;
use Exporter;
use Carp;
use Data::Tools;
use Date::Calc qw(:all);
use Time::JulianDay;

our $VERSION = '1.25';

our @ISA    = qw( Exporter );
our @EXPORT = qw(

                unix_time_diff_in_words
                unix_time_diff_in_words_relative
    
                julian_date_diff_in_words
                julian_date_diff_in_words_relative

                get_local_time_only
                get_local_julian_day
                get_local_year
                get_year_month_days

                julian_date_from_utime
                julian_date_add_ymd
                julian_date_to_ymd
                julian_date_from_ymd

                julian_date_goto_first_dom
                julian_date_goto_last_dom
                julian_date_get_dow
                julian_date_month_days_ym
                julian_date_month_days
                julian_date_get_dow

                utime_from_julian_date
                utime_from_ymdhms
                utime_to_ymdhms
                
                utime_split_to_jdt
                utime_split_to_utt
                utime_join_jdt
                utime_join_utt

                utime_add_ymdhms
                utime_add_ymd
                utime_add_hms

                utime_month_days

                utime_goto_midnight
                utime_goto_first_dom
                utime_goto_last_dom
                utime_goto_first_doy
                utime_goto_last_doy
                utime_get_dow
                utime_get_moy
                
                );

our %EXPORT_TAGS = (
                   
                   'all'  => \@EXPORT,
                   'none' => [],
                   
                   );

##############################################################################

sub unix_time_diff_in_words
{
  my $utd = abs( int( shift() ) ); # absolute difference in seconds

  if( $utd < 1 )
    {
    return "now";
    }
  if( $utd < 60   ) # less than 1 minute
    {
    my $ss = str_countable( $utd, "second", "seconds" );
    return "$utd $ss";
    };
  if( $utd < 60*60 ) # less than 1 hour
    {
    my $m  = int( $utd / 60 );
    my $ms = str_countable( $m, "minute", "minutes" );
    return "$m $ms";
    };
  if( $utd < 2*24*60*60 ) # less than 2 days (48 hours)
    {
    my $h = int( $utd / ( 60 * 60 ) );
    my $m = int( $utd % ( 60 * 60 ) / 60 );
    my $hs = str_countable( $h, "hour",   "hours"   );
    my $ms = str_countable( $m, "minute", "minutes" );
    return "$h $hs, $m $ms";
    };
  if( $utd < 7*24*60*60 ) # less than 1 week (168 hours)
    {
    my $d  = int( $utd / ( 24 * 60 * 60 ) );
    my $h  = int( $utd % ( 24 * 60 * 60 ) / ( 60 * 60 ) );
    my $ds = str_countable( $d, "day",    "days"    );
    my $hs = str_countable( $h, "hour",   "hours"   );
    return "$d $ds, $h $hs";
    };
  if( $utd < 60*24*60*60 ) # less than 2 months
    {
    my $d  = int( $utd / ( 24 * 60 * 60 ) );
    my $ds = str_countable( $d, "day",    "days"    );
    return "$d $ds";
    };
  if( 42 ) # more than 2 months
    {
    my $m  = int( $utd / ( 30*24*60*60 ) ); # "month" is approximated to 30 days
    my $ms = str_countable( $m, "month", "months" );
    return "$m $ms";
    }
}

sub unix_time_diff_in_words_relative
{
  my $utd = int( shift() ); # relative difference in seconds

  my $uts = unix_time_diff_in_words( $utd );

  if( $utd < 0 )
    {
    return "in $uts";
    }
  elsif( $utd > 0 )
    {
    return "before $uts";
    }
  else
    {
    return $uts;
    }
}

##############################################################################

sub julian_date_diff_in_words
{
  my $jdd  = abs( int( shift() ) ); # absolute difference in days

  if( $jdd < 90 )
    {
    my $d  = int( $jdd );
    my $ds = str_countable( $d, "day", "days" );
    return "$d $ds";
    }
  if( 42 )
    {
    my $m  = int( $jdd / 30 );
    my $ms = str_countable( $m, "month", "months" );
    return "$m $ms";
    };
}

sub julian_date_diff_in_words_relative
{
  my $jdd = int( shift() ); # relative difference in days

  if( $jdd == 0 )
    {
    return "today";
    }
  if( $jdd == -1 )
    {
    return "tomorrow";
    }
  if( $jdd == +1 )
    {
    return "yesterday";
    }

  my $jds = julian_date_diff_in_words( $jdd );
  if( $jdd < 0 )
    {
    return "in $jds";
    }
  elsif( $jdd > 0 )
    {
    return "before $jds";
    }
  else
    {
    return $jds;
    }
}

##############################################################################

# returns time of the day (in the current day only)
sub get_local_time_only
{
  my ( $s, $m, $h ) = localtime( shift() || time() );
  return $h*60*60 + $m*60 + $s;
}

sub get_local_julian_day
{
  return local_julian_day( time() );
}

sub get_local_year
{
   my ( $y ) = inverse_julian_day( local_date() );
   return $y;
}

sub get_year_month_days
{
  return Days_in_Month( @_ );
}

# return julian date, moved with positive or negative deltas ( y, m, d ) 
sub julian_date_add_ymd
{
  my $wd = shift; # original/work date
  my $dy = shift; # add delta year
  my $dm = shift; # add delta month
  my $dd = shift; # add delta day

  my ( $y, $m, $d ) = inverse_julian_day( $wd );

  ( $y, $m, $d ) = Add_Delta_YMD( $y, $m, $d, $dy, $dm, $dd );

  $wd = julian_day( $y, $m, $d );

  return $wd;

}

# return ( year, month, day ) from julian date
sub julian_date_to_ymd
{
  my $wd = shift; # original/work date

  my ( $y, $m, $d ) = inverse_julian_day( $wd );
  return ( $y, $m, $d );
}

# return julian date from ( year, month, day )
sub julian_date_from_ymd
{
  my $y = shift; # set year
  my $m = shift; # set month
  my $d = shift; # set day

  my $wd = julian_day( $y, $m, $d );
  return $wd;
}

# return julian date, moved to the first day of its month
sub julian_date_goto_first_dom
{
  my $wd = shift; # original/work date

  my ( $y, $m, $d ) = julian_date_to_ymd( $wd );
  return julian_date_from_ymd( $y, $m, 1 );
}

# return julian date, moved to the last day of its month
sub julian_date_goto_last_dom
{
  my $wd = shift; # original/work date

  my ( $y, $m, $d ) = julian_date_to_ymd( $wd );
  return julian_date_from_ymd( $y, $m, Days_in_Month( $y, $m ) );
}

# return day of the week, for julian date -- 0 Sun .. 6 Sat
sub julian_date_get_dow
{
  my $d = shift; # original date

  return day_of_week( $d );
}

# return month days count for given ( year, month ) (not strictly julian_ namespace)
sub julian_date_month_days_ym
{
  my $y = shift; # set year
  my $m = shift; # set month

  return Days_in_Month( $y, $m );
}

# return month days count for given julian date
sub julian_date_month_days
{
  my $d = shift;

  return Days_in_Month( ( julian_date_to_ymd( $d ) )[0,1] );
}

##############################################################################

sub utime_from_julian_date
{
  my ( $year, $month, $day ) = inverse_julian_day( shift() );
  return Mktime( $year, $month, $day, 0, 0, 0 );
}

sub utime_from_ymdhms
{
  my @args = ( @_, 0, 0, 0, );
  return Mktime( @args[ 0 .. 5 ] );
}

sub utime_to_ymdhms
{
  return Localtime(shift());
}

# returns local julian day and time from unix time
sub utime_split_to_jdt
{
  my ($year,$month,$day,  $hour,$min,$sec,  $doy,$dow,$dst) = Localtime(shift());
  my $jd = julian_day( $year, $month, $day );
  my $tt = $hour * 60 * 60 + $min * 60 + $sec;
  return ( $jd, $tt );
}

# returns unix time of the 00:00 of the day given as utime and time from unix time
sub utime_split_to_utt
{
  my ($year,$month,$day,  $hour,$min,$sec,  $doy,$dow,$dst) = Localtime(shift());
  
  my $ut = Mktime( $year, $month, $day, 0, 0, 0 );
  my $tt = $hour * 60 * 60 + $min * 60 + $sec;
  return ( $ut, $tt );
}

# joins julian date and day time only into unix time
sub utime_join_jdt
{
  my $jd = shift;
  my $tt = shift;
  
  return utime_from_julian_date( $jd ) + $tt;
}

# joins unix time of the day midnight and day time only into unix time
# mostly redundant but exists for completeness
sub utime_join_utt
{
  return shift() + shift();
}

sub utime_add_ymdhms
{
  my $tt = shift;
  my ( $dy, $do, $dd, $dh, $dm, $ds ) = @_;
  
  my ( $year, $month, $day, $hour, $min, $sec ) = Localtime( $tt );
  
  ( $year, $month, $day, $hour, $min, $sec ) =
      Add_Delta_YMDHMS( $year, $month, $day, $hour, $min, $sec,
                        $dy,   $do,    $dd,  $dh,   $dm,  $ds  );

  return Mktime( $year, $month, $day, $hour, $min, $sec );
}

sub utime_add_ymd
{
  my $tt = shift;
  my ( $dy, $do, $dd ) = @_;
  return utime_add_ymdhms( $tt, $dy, $do, $dd, 0, 0, 0 );
}

sub utime_add_hms
{
  my $tt = shift;
  my ( $dh, $dm, $ds ) = @_;
  return utime_add_ymdhms( $tt, 0, 0, 0, $dh, $dm, $ds );
}

sub utime_month_days
{
  return get_year_month_days( ( utime_to_ymdhms( shift() ) )[ 0, 1 ] );
}

sub utime_goto_midnight
{
  return ( utime_split_to_utt( shift() ) )[0];
}

sub utime_goto_first_dom
{
  my $tt = shift;
  my ( $year, $month, $day, $hour, $min, $sec ) = Localtime( $tt );
  return Mktime( $year, $month, 1, 0, 0, 0 );
}

sub utime_goto_last_dom
{
  my $tt = shift;
  my ( $year, $month, $day, $hour, $min, $sec ) = Localtime( $tt );
  return Mktime( $year, $month, Days_in_Month( $year, $month ), 0, 0, 0 );
}

sub utime_goto_first_doy
{
  my $tt = shift;
  my ( $year, $month, $day, $hour, $min, $sec ) = Localtime( $tt );
  return Mktime( $year, 1, 1, 0, 0, 0 );
}

sub utime_goto_last_doy
{
  my $tt = shift;
  my ( $year, $month, $day, $hour, $min, $sec ) = Localtime( $tt );
  return Mktime( $year, 12, 31, 0, 0, 0 );
}

sub utime_get_dow
{
  my $tt = shift;
  my ( $year, $month, $day, $hour, $min, $sec ) = Localtime( $tt );
  return Day_of_Week( $year, $month, $day );
}

sub utime_get_moy
{
  my $tt = shift;
  my ( $year, $month, $day, $hour, $min, $sec ) = Localtime( $tt );
  return $month;
}

##############################################################################

=pod


=head1 NAME

  Data::Tools::Time provides set of basic functions for time processing.

=head1 SYNOPSIS

  use Data::Tools::Time qw( :all );  # import all functions
  use Data::Tools::Time;             # the same as :all :) 
  use Data::Tools::Time qw( :none ); # do not import anything

  # --------------------------------------------------------------------------

  my $time_diff_str     = unix_time_diff_in_words( $time1 - $time2 );
  my $time_diff_str_rel = unix_time_diff_in_words_relative( $time1 - $time2 );

  # --------------------------------------------------------------------------
    
  my $date_diff_str     = julian_date_diff_in_words( $date1 - $date2 );
  my $date_diff_str_rel = julian_date_diff_in_words_relative( $date1 - $date2 );

  # --------------------------------------------------------------------------

  # return seconds after last midnight, i.e. current day time
  my $seconds_in_the_current_day = get_local_time_only()
  
  # returns current julian day
  my $jd = get_local_julian_day()
  
  # returns current year
  my $year = get_local_year()
  
  # gets current julian date, needs Time::JulianDay
  my $jd = local_julian_day( time() );
  # or
  my $jd = get_local_julian_day();

  # move current julian date to year ago, one month ahead and 2 days ahead
  $jd = julian_date_add_ymd( $jd, -1, 1, 2 );

  # get year, month and day from julian date
  my ( $y, $m, $d ) = julian_date_to_ymd( $jd );

  # get julian date from year, month and day
  $jd = julian_date_from_ymd( $y, $m, $d );

  # move julian date ($jd) to the first day of its current month
  $jd = julian_date_goto_first_dom( $jd );

  # move julian date ($jd) to the last day of its current month
  $jd = julian_date_goto_last_dom( $jd );

  # get day of week for given julian date ( 0 => Mon .. 6 => Sun )
  my $dow = julian_date_get_dow( $jd );
  print( ( qw( Mon Tue Wed Thu Fri Sat Sun ) )[ $dow ] . "\n" );

  # get month days count for the given julian date's month
  my $mdays = julian_date_month_days( $jd );

  # get month days count for the given year and month
  my $mdays = julian_date_month_days_ym( $y, $m );

=head1 FUNCTIONS

=head2 unix_time_diff_in_words( $unix_time_diff )

Returns human-friendly text for the given time difference (in seconds).
This function returns absolute difference text, for relative 
(before/after/ago/in) see unix_time_diff_in_words_relative().

=head2 unix_time_diff_in_words_relative( $unix_time_diff )

Same as unix_time_diff_in_words() but returns relative text
(i.e. with before/after/ago/in)

=head2 julian_date_diff_in_words( $julian_date_diff );

Returns human-friendly text for the given date difference (in days).
This function returns absolute difference text, for relative 
(before/after/ago/in) see julian_day_diff_in_words_relative().

=head2 julian_date_diff_in_words_relative( $julian_date_diff );

Same as julian_date_diff_in_words() but returns relative text
(i.e. with before/after/ago/in)

=head1 TODO

  * support for language-dependent wording (before/ago)
  * support for user-defined thresholds (48 hours, 2 months, etc.)

=head1 REQUIRED MODULES

Data::Tools::Time uses:

  * Data::Tools (from the same package)
  * Date::Calc
  * Time::JulianDay

=head1 TEXT TRANSLATION NOTES

time/date difference wording functions does not have translation functions
and return only english text. This is intentional since the goal is to keep
the translation mess away but still allow simple (yet bit strange) 
way to translate the result strings with regexp and language hash:
  
  my $time_diff_str_rel = unix_time_diff_in_words_relative( $time1 - $time2 );
  
  my %TRANS = (
              'now'       => 'sega',
              'today'     => 'dnes',
              'tomorrow'  => 'utre',
              'yesterday' => 'vchera',
              'in'        => 'sled',
              'before'    => 'predi',
              'year'      => 'godina',
              'years'     => 'godini',
              'month'     => 'mesec',
              'months'    => 'meseca',
              'day'       => 'den',
              'days'      => 'dni',
              'hour'      => 'chas',
              'hours'     => 'chasa',
              'minute'    => 'minuta',
              'minutes'   => 'minuti',
              'second'    => 'sekunda',
              'seconds'   => 'sekundi',
              );
              
  $time_diff_str_rel =~ s/([a-z]+)/$TRANS{ lc $1 } || $1/ge;

I know this is no good for longer sentences but works fine in this case.

=head1 GITHUB REPOSITORY

  git@github.com:cade-vs/perl-data-tools.git
  
  git clone git://github.com/cade-vs/perl-data-tools.git
  
=head1 AUTHOR

  Vladi Belperchinov-Shabanski "Cade"

  <cade@bis.bg> <cade@biscom.net> <cade@datamax.bg> <cade@cpan.org>

  http://cade.datamax.bg


=cut

##############################################################################
1;
